File: li/cil/oc/api/detail/FileSystemAPI.scala
package li.cil.oc.api.detail

import li.cil.oc.api.FileSystem
import li.cil.oc.api.network.Node
import dan200.computer.api.{IMount, IWritableMount}

/** Avoids reflection structural types would induce. */
trait FileSystemAPI {
  def fromClass(clazz: Class[_], domain: String, root: String): Option[FileSystem]

  def fromSaveDirectory(root: String, capacity: Long, buffered: Boolean): Option[FileSystem]

  def fromMemory(capacity: Long): Option[FileSystem]

  def fromComputerCraft(mount: IMount): Option[FileSystem]

  def fromComputerCraft(mount: IWritableMount): Option[FileSystem]

  def asNode(fs: FileSystem): Option[Node]
}

File: li/cil/oc/api/detail/FileSystemAPI.java
package li.cil.oc.api.detail;

import dan200.computer.api.IMount;
import dan200.computer.api.IWritableMount;
import li.cil.oc.api.fs.FileSystem;
import li.cil.oc.api.network.ManagedEnvironment;

public interface FileSystemAPI {
    FileSystem fromClass(Class<?> clazz, String domain, String root);

    FileSystem fromSaveDirectory(String root, long capacity, boolean buffered);

    FileSystem fromMemory(long capacity);

    FileSystem fromComputerCraft(IMount mount);

    FileSystem fromComputerCraft(IWritableMount mount);

    ManagedEnvironment asManagedEnvironment(FileSystem fs, String label);
}